### PR TITLE
Store CSP delivered via meta tag as a valid HTTP header.

### PR DESCRIPTION
**Pull request by rreno**
#### 0445ac553799b27c80ebc292d372e0663ad70b8a
<pre>
Store CSP delivered via meta tag as a valid HTTP header.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249596">https://bugs.webkit.org/show_bug.cgi?id=249596</a>
rdar://103170891

Reviewed by Brent Fulgham.

A CSP delivered via a meta tag could have invalid HTTP header values in it. Take for example this:

&lt;meta http-equiv=&quot;Content-Security-Policy&quot; content=&quot;
    default-src &apos;none&apos;;
    script-src &apos;self&apos;;
    img-src &apos;self&apos;&quot;&gt;

The value of the CSP header that the ContentSecurityPolicyDirectiveList will get will be the raw
string including whitespace and most importantly newline characters. These newline characters are
invalid characters in an HTTP header[0].

The parsing algorithm for CSP handles this appropriately and creates a valid CSP for the document. However,
if a script in the document then creates blob URLs which are navigated to or otherwise fetched, the Network
process will return a ResourceResponse object with a Content-Security-Policy header that contains the newlines.
This is caught by the ResourceResponseBase::containsInvalidHTTPHeaders function which causes the fetch to fail.

To combat this we can simply strip the newline characters from the meta-delivered CSP and store the policy as a
valid HTTP header.

[0] <a href="https://fetch.spec.whatwg.org/#header-value">https://fetch.spec.whatwg.org/#header-value</a>

* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::parse):

Canonical link: <a href="https://commits.webkit.org/258110@main">https://commits.webkit.org/258110@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/75e89dec3b1f71528baed4d0e66e8076f817cd6c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110241 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104924 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/960 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93348 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108081 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106719 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34945 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22991 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3793 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9905 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44010 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5565 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5568 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp [hunk 1]
```diff
@@ -444,7 +445,15 @@ const ContentSecurityPolicyDirective* ContentSecurityPolicyDirectiveList::violat
 //
 void ContentSecurityPolicyDirectiveList::parse(const String& policy, ContentSecurityPolicy::PolicyFrom policyFrom)
 {
-    m_header = policy;
+    // A meta tag delievered CSP could contain invalid HTTP header values depending on how it was formatted in the document.
+    // We want to store the CSP as a valid HTTP header for e.g. blob URL inheritance.
+    if (policyFrom == ContentSecurityPolicy::PolicyFrom::HTTPEquivMeta) {
+        m_header = stripLeadingAndTrailingHTTPSpaces(policy).removeCharacters([](auto c) {
+            return c == 0x00 || c == '\r' || c == '\n';
```
Comment (brentfulgham) [inline Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp]: It seems like we should have a test for this (or perhaps WPT has one, already?).

**Review by brentfulgham — APPROVED** [review]
Looks good -- can you please follow-up with a test case (perhaps in a second PR?)

File: Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp (hunk 1 above)
Comment (rreno) [reply Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp]: Will do in a follow-up!

**Comment by rreno** [comment]
When EWS is green I'm going to push an amended commit message with some typo fixes and the fetch spec reference I meant to include.

**Pull request merged**